### ISSUE TITLE
Recursive dispatch

### ISSUE THREAD

**Issue by HeMet** [ReSwift/ReSwift#74]
Currently subscribers are called just before default dispatcher returns its result. If one of subscribers produce an action then we get recursive dispatch.

Example:

```swift
struct NewStringValue: Action {
    let value: String
}

class StringReducer: Reducer {
    func handleAction(action: Action, state: String?) -> String {
        let newString = action as! NewStringValue
        return newString.value
    }
}

class StringSubscriber: StoreSubscriber {
    var count = 0
    func newState(state: String) {
        if count < 3 {
            count += 1
            stringStore.dispatch(NewStringValue(value: state + "@"))
        }
    }
}

// I modified Middleware signature slightly while I have been experimenting on ReSwift.
let loggingMiddleware: Middleware = { store in
    return { next in
        return { action in
            print("action: \(action)")
            let result = next(action)
            print(store?.optState ?? "state is not hydrated yet")
            return result
        }
    }
}

extension String: StateType { }

let stringStore = Store(reducer: StringReducer(), state: "", middleware: [loggingMiddleware])
let subscriber = StringSubscriber()
stringStore.subscribe(subscriber)
```

Output:
```
action: NewStringValue(value: "@")
action: NewStringValue(value: "@@")
action: NewStringValue(value: "@@@")
@@@
@@@
@@@
```

Is it by design? Workaround is to dispatch actions asynchronously.

**Comment by Ben-G** [comment]
Hi @HeMet! I haven't yet have time to investigate the full issue. But I can say that everything works expected when not using a middleware to log the store state.

I used your test case and investigated the state value within the `newState` method and it returns the expected sequence:

```
@
@@
@@@
```

When I have some more time, I will look into whether there's an issue with the way the middleware gets a reference to the store's current state.

**Comment by Ben-G** [comment]
Ah, I had a few more moments to look into this; the issue lies in the implementation of the middleware:

```
let loggingMiddleware: Middleware = { store in
    return { next in
        return { action in
            print("action: \(action)")
            let result = next(action)
            print(store?.optState ?? "state is not hydrated yet")
            return result
        }
    }
}
```

Calling the next middleware should be the last step of your middleware:

```
let loggingMiddleware: Middleware = { store in
    return { next in
        return { action in
            print("action: \(action)")
            print(store?.optState ?? "state is not hydrated yet")
            let result = next(action)
            return result
        }
    }
}
```

This should yield the correct sequence. However, I will still look into whether the behavior you are seeing is expected or not.

Cheers!

**Comment by agentk** [comment]
This is also an issue noted with Redux: https://github.com/rackt/redux/issues/1221
> The behavior is intentional; subscribers aren't guaranteed to get every state update, just the most recent one. If you have ideas how to optimize this please send a PR!

Changing the order of middleware code is one way to fix it. It's more of a an issue with subscribers though. It's best practice not to dispatch new events from a subscription callback. If you must, they can be triggered for the the next loop cycle using `dispatch_after` or cached in queue for dispatch.

It is possible to extend the `dispatch in progress` guard in the store, although that would be preventing borderline, but sometimes handy behaviour.

tl;dr. Behavour is correct, although unexpected. Known gotcha. Middleware can only see the latest state. Sometimes required for certain state changes. Should be kept minimal as it's easy to introduce side-effects.

**Comment by Ben-G** [comment]
Thanks @agentk for linking to the issue on redux. I had a little bit more time to think about this specific case and the behavior does make sense. 

Since you only have a single middleware and forward to the `next` middleware before printing the current state, you are executing the `_defaultDispatch` before you're getting the current state of the store. 

This means that control to your middleware is only returned once `_defaultDispatch` returns. That however happens **after** state updates are sent to subscribers. Since you're dispatching in the subscriber, `_defaultDispatch` will be invoked another time, before the original call returned.

Only once you stop dispatching from the subscriber, the original `_defaultDispatch` will return and your middleware will get a chance to print the current state.

Here are the two relevant lines in `_defaultDispatch`:
```
public func _defaultDispatch(action: Action) -> Any {
        if isDispatching {
            // Use Obj-C exception since throwing of exceptions can be verified through tests
            NSException.raise("SwiftFlow:IllegalDispatchFromReducer", format: "Reducers may not " +
                "dispatch actions.", arguments: getVaList(["nil"]))
        }

        isDispatching = true
        let newState = reducer._handleAction(action, state: state) as! State
        isDispatching = false

        // subscribers are notified here:
        state = newState

        // control is returned to your middleware here:
        return action
    }
```

This means the behavior makes sense. As a rule of thumb your last middleware should probably call `next` (which is `_defaultDispatch`) as the last call in the middleware implementation to avoid this scenario. I also cannot think of a use case where you would benefit from not doing that.

@HeMet I hope this clarifies your question! I'll leave this open for discussion for a little longer.

**Comment by HeMet** [comment]
@Ben-G Thanks for explanation!

> Calling the next middleware should be the last step of your middleware:

In this case output is unexpected too:
```
action: NewStringValue(value: "@")

action: NewStringValue(value: "@@")
@
action: NewStringValue(value: "@@@")
@@
```
Since I expected even lines to describe new state after an action dispatched:
```
action: NewStringValue(value: "@")
@
action: NewStringValue(value: "@@")
@@
action: NewStringValue(value: "@@@")
@@@
```
> Since you only have a single middleware and forward to...

That was the conclusion I reached after I have looked into the implementation.

@agentk Thanks for the link and explanation!
```
@@@
@@@
@@@
```
Actually I was so focused on the sequence issue that overlooked this one :smile: So two gotchas instead of one.

**Comment by Ben-G** [comment]
@HeMet This is actually not unexpected:

```
action: NewStringValue(value: "@")

action: NewStringValue(value: "@@")
@
action: NewStringValue(value: "@@@")
@@
```

The state update happens after your middleware is called, so this behavior is expected.

**Comment by HeMet** [comment]
> This is actually not unexpected
Yeah, I understand it now. I mean It was unexpected to me in time I wrote that code.

**Comment by Ben-G** [comment]
@HeMet Nice! I definitely see how this is not necessarily intuitive at first!
